### PR TITLE
[spirv] Add location to error messages and unify.

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 1]
```diff
@@ -2638,7 +2652,8 @@ SPIRVEmitter::processBufferTextureLoad(const CXXMemberCallExpr *expr) {
     return processBufferTextureLoad(object, coordinate, constOffset, varOffset,
                                     lod);
   }
-  emitError("Load() is not implemented for the given object type.");
+  emitError("Load() is not implemented for the given object type.",
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Load() of the given object type unimplemented

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 2]
```diff
@@ -2656,7 +2671,8 @@ uint32_t SPIRVEmitter::processGetDimensions(const CXXMemberCallExpr *expr) {
              TypeTranslator::isConsumeStructuredBuffer(objectType)) {
     return processByteAddressBufferStructuredBufferGetDimensions(expr);
   } else {
-    emitError("GetDimensions not implmented for the given type yet.");
+    emitError("GetDimensions not implmented for the given type yet.",
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: GetDimensions() of the given object type unimplemented

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 3]
```diff
@@ -3225,8 +3242,8 @@ SpirvEvalInfo SPIRVEmitter::processBinaryOp(const Expr *lhs, const Expr *rhs,
     break;
   }
 
-  emitError("BinaryOperator '%0' is not supported yet.")
-      << BinaryOperator::getOpcodeStr(opcode);
+  emitError("BinaryOperator '%0' is not supported yet.", lhs->getExprLoc())
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: binary operator '%0' unimplemented

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 4]
```diff
@@ -3791,8 +3814,9 @@ SPIRVEmitter::processMatrixBinaryOp(const Expr *lhs, const Expr *rhs,
     break;
   }
 
-  emitError("BinaryOperator '%0' for matrices not supported yet")
-      << BinaryOperator::getOpcodeStr(opcode);
+  emitError("BinaryOperator '%0' for matrices not supported yet",
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: binary operator '%0' over matrix type unimplemented

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 5]
```diff
@@ -3904,17 +3928,18 @@ uint32_t SPIRVEmitter::castToInt(const uint32_t fromVal, QualType fromType,
     } else if (isUintOrVecOfUintType(toIntType)) {
       return theBuilder.createUnaryOp(spv::Op::OpConvertFToU, intType, fromVal);
     } else {
-      emitError("unimplemented casting to integer from floating point");
+      emitError("unimplemented casting to integer from floating point", srcLoc);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: casting from floating point to integer unimplemented

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 6]
```diff
@@ -3904,17 +3928,18 @@ uint32_t SPIRVEmitter::castToInt(const uint32_t fromVal, QualType fromType,
...
     }
   } else {
-    emitError("unimplemented casting to integer");
+    emitError("unimplemented casting to integer", srcLoc);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: casting to integer unimplemented

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 7]
```diff
@@ -3935,11 +3960,11 @@ uint32_t SPIRVEmitter::castToFloat(const uint32_t fromVal, QualType fromType,
   }
 
   if (isFloatOrVecOfFloatType(fromType)) {
-    emitError("casting between different fp bitwidth unimplemented");
+    emitError("casting between different fp bitwidth unimplemented", srcLoc);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: s/fp/floating point/

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 8]
```diff
@@ -3935,11 +3960,11 @@ uint32_t SPIRVEmitter::castToFloat(const uint32_t fromVal, QualType fromType,
...
     return 0;
   }
 
-  emitError("unimplemented casting to floating point");
+  emitError("unimplemented casting to floating point", srcLoc);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: casting to floating point unimplemented

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 9]
```diff
@@ -4153,7 +4178,8 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
     INTRINSIC_OP_CASE(sqrt, Sqrt, true);
     INTRINSIC_OP_CASE(trunc, Trunc, true);
   default:
-    emitError("Intrinsic function '%0' not yet implemented.")
+    emitError("intrinsic function %0 not yet implemented",
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: intrinsic '%0' function unimplemented

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 10]
```diff
@@ -4363,7 +4391,8 @@ uint32_t SPIRVEmitter::processIntrinsicModf(const CallExpr *callExpr) {
     }
   }
 
-  emitError("Unknown argument type passed to Modf function.");
+  emitError("unknown argument type passed to Modf function",
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: invalid argument type passed to Modf intrinsic function?

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 11]
```diff
@@ -4484,7 +4513,8 @@ uint32_t SPIRVEmitter::processIntrinsicFrexp(const CallExpr *callExpr) {
     }
   }
 
-  emitError("Unknown argument type passed to Frexp function.");
+  emitError("unknown argument type passed to Frexp function",
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: invalid argument type passed to Frexp intrinsic function?

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 12]
```diff
@@ -4730,7 +4760,8 @@ uint32_t SPIRVEmitter::processIntrinsicMul(const CallExpr *callExpr) {
     }
   }
 
-  emitError("Unsupported arguments passed to mul() function.");
+  emitError("unsupported arguments passed to mul() function",
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: invalid argument type passed to mul intrinsic function?

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 13]
```diff
@@ -4873,7 +4904,8 @@ uint32_t SPIRVEmitter::processIntrinsicAllOrAny(const CallExpr *callExpr,
                                     &matColCount)) {
       if (!elemType->isFloatingType()) {
         emitError("'all' and 'any' currently do not take non-floating point "
-                  "matrices as argument.");
+                  "matrices as argument",
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: non-floating-point arguments in all/any intrinsic function unimplemented

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 14]
```diff
@@ -4539,7 +4569,7 @@ uint32_t SPIRVEmitter::processIntrinsicClip(const CallExpr *callExpr) {
         theBuilder.createCompositeConstruct(boolRowType, cmpResults);
     condition = theBuilder.createUnaryOp(spv::Op::OpAny, boolType, results);
   } else {
-    emitError("Invalid type passed to clip function.");
+    emitError("invalid type passed to clip function", callExpr->getExprLoc());
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: invalid argument type passed to clip intrinsic function?

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 15]
```diff
@@ -5089,7 +5121,8 @@ uint32_t SPIRVEmitter::processIntrinsicSaturate(const CallExpr *callExpr) {
     return processEachVectorInMatrix(arg, argId, actOnEachVec);
   }
 
-  emitError("Invalid argument type passed to saturate().");
+  emitError("invalid argument type passed to saturate()",
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: ... to saturate intrinsic function?

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 16]
```diff
@@ -5396,7 +5430,7 @@ uint32_t SPIRVEmitter::translateAPValue(const APValue &value,
     return theBuilder.getConstantComposite(vecType, elements);
   }
 
-  emitError("APValue of type '%0' is not supported yet.") << value.getKind();
+  emitError("APValue of type %0 is not supported yet", {}) << value.getKind();
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: APValue of type %0 unimplemented

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 17]
```diff
@@ -5415,7 +5449,7 @@ uint32_t SPIRVEmitter::translateAPInt(const llvm::APInt &intValue,
           static_cast<uint32_t>(intValue.getZExtValue()));
   }
 
-  emitError("APInt for target bitwidth '%0' is not supported yet.")
+  emitError("APInt for target bitwidth %0 is not supported yet", {})
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: s/is not supported yet/unimplemented/

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 18]
```diff
@@ -5435,7 +5469,7 @@ uint32_t SPIRVEmitter::translateAPFloat(const llvm::APFloat &floatValue,
     break;
   }
 
-  emitError("APFloat for target bitwidth '%0' is not supported yet.")
+  emitError("APFloat for target bitwidth %0 is not supported yet", {})
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: s/is not supported yet/unimplemented/

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 19]
```diff
@@ -6025,7 +6061,8 @@ void SPIRVEmitter::discoverAllCaseStmtInSwitchStmt(
     auto bitWidth = astContext.getIntWidth(caseExpr->getType());
     if (bitWidth != 32)
       emitError("Switch statement translation currently only supports 32-bit "
-                "integer case values.");
+                "integer case values",
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: non-32bit integer case value in switch statement unimplemented

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 20]
```diff
@@ -149,7 +149,7 @@ uint32_t TypeTranslator::translateType(QualType type, LayoutRule rule,
         case BuiltinType::Double:
           return theBuilder.getFloat64Type();
         default:
-          emitError("Primitive type '%0' is not supported yet.")
+          emitError("primitive type %0 is not supported yet")
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: s/is not supported yet/unimplemented/

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 21]
```diff
@@ -259,7 +259,7 @@ uint32_t TypeTranslator::translateType(QualType type, LayoutRule rule,
                                    decorations);
   }
 
-  emitError("Type '%0' is not supported yet.") << type->getTypeClassName();
+  emitError("type %0 is not supported yet") << type->getTypeClassName();
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: s/is not supported yet/unimplemented/

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 22]
```diff
@@ -289,13 +289,14 @@ uint32_t TypeTranslator::getGlPerVertexStruct(uint32_t clipArraySize,
...
 
-  decorations.push_back(Decoration::getBuiltIn(ctx, spv::BuiltIn::Position, 0));
-  decorations.push_back(
-      Decoration::getBuiltIn(ctx, spv::BuiltIn::PointSize, 1));
-  decorations.push_back(
-      Decoration::getBuiltIn(ctx, spv::BuiltIn::ClipDistance, 2));
-  decorations.push_back(
-      Decoration::getBuiltIn(ctx, spv::BuiltIn::CullDistance, 3));
+  decorations.push_back(Decoration::getBuiltIn(ctx, spv::BuiltIn::Position,
+                                               llvm::Optional<uint32_t>(0)));
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: Bad rebasing? `llvm::Optional<..>()` should already be removed.

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 23]
```diff
@@ -934,7 +935,7 @@ TypeTranslator::getAlignmentAndSize(QualType type, LayoutRule rule,
     return {alignment, size};
   }
 
-  emitError("Type '%0' is not supported yet.") << type->getTypeClassName();
+  emitError("type %0 is not supported yet") << type->getTypeClassName();
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: s/is not supported yet/unimplemented/

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 24]
```diff
@@ -841,7 +842,7 @@ TypeTranslator::getAlignmentAndSize(QualType type, LayoutRule rule,
         case BuiltinType::Float:
           return {4, 4};
         default:
-          emitError("Primitive type '%0' is not supported yet.")
+          emitError("primitive type %0 is not supported yet")
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: s/is not supported yet/unimplemented/

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 25]
```diff
@@ -765,7 +766,7 @@ TypeTranslator::translateSampledTypeToImageFormat(QualType sampledType) {
       }
     }
   }
-  emitError("Unimplemented resource result type was used.");
+  emitError("unimplemented resource result type was used");
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: resource type %0 unimplemented ... << sampledType

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 26]
```diff
@@ -1128,14 +1130,16 @@ void DeclResultIdMapper::decoratePSInterpolationMode(const DeclaratorDecl *decl,
...
+                                                 SourceLocation srcLoc) {
   using spv::BuiltIn;
   const auto sigPoint = stageVar->getSigPoint();
   const auto semanticKind = stageVar->getSemantic()->GetKind();
   const auto sigPointKind = sigPoint->GetKind();
   const uint32_t type = stageVar->getSpirvTypeId();
 
-  spv::StorageClass sc = getStorageClassForSigPoint(sigPoint);
+  spv::StorageClass sc = getStorageClassForSigPoint(
+      sigPoint, stageVar->getSemantic()->GetName(), srcLoc);
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Done

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 27]
```diff
@@ -1189,7 +1193,7 @@ uint32_t DeclResultIdMapper::createSpirvStageVar(StageVar *stageVar,
     case hlsl::SigPoint::Kind::PSIn:
       return theBuilder.addStageIOVar(type, sc, name.str());
     default:
-      emitError("semantic InstanceID for SigPoint %0 unimplemented yet")
+      emitError("semantic InstanceID for SigPoint %0 unimplemented yet", srcLoc)
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Done

[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Done

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 28]
```diff
@@ -1243,7 +1247,8 @@ uint32_t DeclResultIdMapper::createSpirvStageVar(StageVar *stageVar,
       stageVar->setIsSpirvBuiltin();
       return theBuilder.addStageBuiltinVar(type, sc, BuiltIn::FrontFacing);
     default:
-      emitError("semantic IsFrontFace for SigPoint %0 unimplemented yet")
+      emitError("semantic IsFrontFace for SigPoint %0 unimplemented yet",
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Done

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 29]
```diff
@@ -1296,7 +1301,7 @@ uint32_t DeclResultIdMapper::createSpirvStageVar(StageVar *stageVar,
     return theBuilder.addStageBuiltinVar(type, sc, BuiltIn::TessCoord);
   }
   default:
-    emitError("semantic %0 unimplemented yet")
+    emitError("semantic %0 unimplemented yet", srcLoc)
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Done

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 30]
```diff
@@ -1328,7 +1335,8 @@ DeclResultIdMapper::getStorageClassForSigPoint(const hlsl::SigPoint *sigPoint) {
       sc = spv::StorageClass::Input;
       break;
     default:
-      emitError("Found invalid SigPoint kind for a semantic.");
+      emitError("SigPoint kind %0 is invalid for semantic %1", srcLoc)
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: That's right. We should turn these into llvm_unreachable.

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 31]
```diff
@@ -1345,12 +1353,14 @@ DeclResultIdMapper::getStorageClassForSigPoint(const hlsl::SigPoint *sigPoint) {
...
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Done

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 32]
```diff
@@ -1345,12 +1353,14 @@ DeclResultIdMapper::getStorageClassForSigPoint(const hlsl::SigPoint *sigPoint) {
...
+          << sigPoint->GetName() << semanticName;
     }
     break;
   }
   default:
-    emitError("Found invalid SignatureKind for semantic.");
+    emitError("found invalid SignatureKind for semantic %0", srcLoc)
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Done

File: tools/clang/lib/SPIRV/DeclResultIdMapper.h [hunk 33]
```diff
@@ -355,7 +356,9 @@ class DeclResultIdMapper {
 
   /// Returns the proper SPIR-V storage class (Input or Output) for the given
   /// SigPoint.
-  spv::StorageClass getStorageClassForSigPoint(const hlsl::SigPoint *);
+  spv::StorageClass getStorageClassForSigPoint(const hlsl::SigPoint *,
+                                               const char *semanticName,
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.h]: Removed

File: tools/clang/lib/SPIRV/InitListHandler.cpp [hunk 34]
```diff
@@ -185,41 +185,43 @@ bool InitListHandler::tryToSplitConstantArray() {
...
 
   if (type->isStructureType())
     return createInitForStructType(type);
 
   if (type->isConstantArrayType())
-    return createInitForConstantArrayType(type);
+    return createInitForConstantArrayType(type, srcLoc);
 
-  emitError("unimplemented initializer for type %0") << type;
+  emitError("unimplemented initializer for type %0", srcLoc) << type;
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/InitListHandler.cpp]: done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 35]
```diff
@@ -364,7 +364,7 @@ void SPIRVEmitter::doDecl(const Decl *decl) {
     llvm_unreachable("HLSLBufferDecl should not be handled here");
   } else {
     // TODO: Implement handling of other Decl types.
-    emitWarning("Decl type '%0' is not supported yet.")
+    emitWarning("decl type %0 is not supported yet", decl->getLocation())
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 36]
```diff
@@ -406,7 +406,8 @@ void SPIRVEmitter::doStmt(const Stmt *stmt,
   } else if (const auto *attrStmt = dyn_cast<AttributedStmt>(stmt)) {
     doStmt(attrStmt->getSubStmt(), attrStmt->getAttrs());
   } else {
-    emitError("Stmt '%0' is not supported yet.") << stmt->getStmtClassName();
+    emitError("Stmt '%0' is not supported yet.", stmt->getLocStart())
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done.
Stmt class does not have a `getLocation()` method. Only has `getLocStart()` and `getLocEnd()`

[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: I'll also add `getSourceRange()` here.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 37]
```diff
@@ -486,7 +487,8 @@ SpirvEvalInfo SPIRVEmitter::doExpr(const Expr *expr) {
     return curThis;
   }
 
-  emitError("Expr '%0' is not supported yet.") << expr->getStmtClassName();
+  emitError("Expr '%0' is not supported yet.", expr->getExprLoc())
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 38]
```diff
@@ -688,18 +690,20 @@ void SPIRVEmitter::doVarDecl(const VarDecl *decl) {
...
   case attr::HLSLUnroll:
     return spv::LoopControlMask::Unroll;
   case attr::HLSLAllowUAVCondition:
-    emitWarning("Unsupported allow_uav_condition attribute ignored.");
+    emitWarning("unsupported allow_uav_condition attribute ignored",
+                stmt->getLocStart());
     break;
   default:
-    emitError("Found unknown loop attribute.");
+    emitError("found unknown loop attribute", stmt->getLocStart());
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 39]
```diff
@@ -1563,12 +1571,12 @@ SpirvEvalInfo SPIRVEmitter::doCastExpr(const CastExpr *expr) {
...
     } else {
-      return processFlatConversion(toType, subExpr->getType(), doExpr(subExpr));
+      return processFlatConversion(toType, subExpr->getType(), doExpr(subExpr),
+                                   expr->getExprLoc());
     }
   }
   default:
-    emitError("ImplictCast Kind '%0' is not supported yet.",
-              expr->getLocStart())
+    emitError("ImplictCast Kind '%0' is not supported yet.", expr->getExprLoc())
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 40]
```diff
@@ -2088,7 +2100,7 @@ uint32_t SPIRVEmitter::processBufferTextureLoad(const Expr *object,
   } else if (elemType->isUnsignedIntegerType()) {
     elemTypeId = theBuilder.getUint32Type();
   } else {
-    emitError("Unimplemented Buffer/Texture type");
+    emitError("Unimplemented Buffer/Texture type", object->getExprLoc());
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 41]
```diff
@@ -2140,7 +2152,8 @@ uint32_t SPIRVEmitter::processByteAddressBufferLoadStore(
     if (expr->getNumArgs() == 2) {
 
       emitError("Load(in Address, out Status) has not been implemented for "
-                "(RW)ByteAddressBuffer yet.");
+                "(RW)ByteAddressBuffer yet.",
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 42]
```diff
@@ -2215,7 +2228,8 @@ uint32_t SPIRVEmitter::processByteAddressBufferLoadStore(
 SpirvEvalInfo
 SPIRVEmitter::processStructuredBufferLoad(const CXXMemberCallExpr *expr) {
   if (expr->getNumArgs() == 2) {
-    emitError("Load(int, int) unimplemented for (RW)StructuredBuffer");
+    emitError("Load(int, int) unimplemented for (RW)StructuredBuffer",
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 1 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 2 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 3 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 4 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 5 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 6 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 7 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 8 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 9 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 10 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 11 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 14 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 12 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 13 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 15 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 16 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 17 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 18 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 19 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/TypeTranslator.cpp (hunk 20 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: done

File: tools/clang/lib/SPIRV/TypeTranslator.cpp (hunk 21 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: done

File: tools/clang/lib/SPIRV/TypeTranslator.cpp (hunk 22 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: This is odd! fixed.

File: tools/clang/lib/SPIRV/TypeTranslator.cpp (hunk 25 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: Done

File: tools/clang/lib/SPIRV/TypeTranslator.cpp (hunk 24 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: done

File: tools/clang/lib/SPIRV/TypeTranslator.cpp (hunk 23 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: done

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 43]
```diff
@@ -766,7 +765,7 @@ TypeTranslator::translateSampledTypeToImageFormat(QualType sampledType) {
       }
     }
   }
-  emitError("unimplemented resource result type was used");
+  emitError("resource type %0 unimplemented") << sampledType.getAsString();
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: Can actually directly `<< sampledType` here?

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 36 above)
[the remark this answers is not in the document]
Comment (antiagainst) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: I see.

File: tools/clang/lib/SPIRV/TypeTranslator.cpp (hunk 43 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: Not sure... Will there be implicit casting to string? Can address in followup CL

**Pull request merged**